### PR TITLE
Status Summary Overhaul

### LINKED ISSUE #4465: ui: Overhaul of {Node|Store}StatusSummaries structures
In addition to Time Series, the metrics system currently generates routine "StatusSummary" objects for every Node and Store.  These structures can be thought of as a sort of "vertical" time series query, quickly returning the most recent value of several time series relevant to a single node or store.  This is a very efficient way to generate the data tables in our UI.

However, these fields are currently stored as protocol buffers with a sub-optimal message schema; this is not an effective structure, as it is not straightforward to expand them with additional metrics, and there is little benefit to enforcing a rigid schema on something consumed only by the UI.

My suggestions for the overhaul:
1. Remove the rigid schema of these objects; they do not need to contain descriptors or structured MVCC data at the moment, they can simply contain a set of metrics values.
2. Store these objects directly as serialized JSON. These summaries are read directly by the UI, and there is no need to perform any post-processing on the server during a query.

Out of scope for this issiue: We will not attempt to store the Summary objects in a SQL table. They are currently being stored in the KV layer, and this will suffice for the time being.

This is related to #4021

### PR DESCRIPTION
**Pull request by mrtracy**
This commit represents a significant overhaul of the status summary system.
- Store status summaries are no longer individually stored or queryable;
  instead, store summaries are written as part of their containing node's summary.
  This is consistent with our current and planned usage of summaries in our UI.
- Summary messages no longer have a rigid schema of metrics; instead, each
  summary message is a "dictionary" of name/value pairs, thus containing
  information for all metrics tracked by both the node and its stores.

Resolves #4465

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5412)

### CODE REVIEW

**Comment by tamird** [comment]
Reviewed 8 of 13 files at r1.
Review status: 8 of 13 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[server/status/status.proto, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDHaIDma-CLb8nENKKN:-KDHaIDma-CLb8nENKKO:-201212321) ([raw file](https://github.com/cockroachdb/cockroach/blob/2226011332ac5969882f4dcbb29d1848c43c1832/server/status/status.proto#L32)):*
any reason not to actually use a map for these?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

**Comment by mrtracy** [comment]
Review status: 8 of 13 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[server/status/status.proto, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDHaIDma-CLb8nENKKN:-KDK-E1IYbTZ7ZpLoEjg:165166636) ([raw file](https://github.com/cockroachdb/cockroach/blob/2226011332ac5969882f4dcbb29d1848c43c1832/server/status/status.proto#L32)):*
Because I didn't know how. From your suggestion, I assume there is some technique with gogoproto to marshal this structure into a map, but I am not familiar with those custom options.

If this _was_ implemented as a map, it would obviate the need for some of the list->map conversions and slice sorting in the test code. However, it would not have an impact on the complexity of the product code, so I suggest we leave this as a post-beta TODO unless there are other significant problems with the code.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

**Comment by tamird** [comment]
Review status: 8 of 13 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[server/status/status.proto, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDHaIDma-CLb8nENKKN:-KDK4OWoUwuXDAEYysPG:-969843473) ([raw file](https://github.com/cockroachdb/cockroach/blob/2226011332ac5969882f4dcbb29d1848c43c1832/server/status/status.proto#L32)):*
It's not gogoproto magic - protocol buffers natively support maps, see gossip.proto for an example.

This is going to be harder to change post-beta, since it affects protobuf encoding.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

**Comment by tamird** [comment]
Review status: 8 of 13 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[server/status/status.proto, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDHaIDma-CLb8nENKKN:-KDK4geZSTQ-Q-TcO3ti:-730470101) ([raw file](https://github.com/cockroachdb/cockroach/blob/2226011332ac5969882f4dcbb29d1848c43c1832/server/status/status.proto#L32)):*
Addendum: gossip.proto is a proto3 file, but maps are supported in proto2 as well.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

**Comment by mrtracy** [comment]
Review status: 8 of 13 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[server/status/status.proto, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDHaIDma-CLb8nENKKN:-KDK5BaRb1a8HrHA7wnE:2132634500) ([raw file](https://github.com/cockroachdb/cockroach/blob/2226011332ac5969882f4dcbb29d1848c43c1832/server/status/status.proto#L32)):*
Ugh, how on earth did I not know about this. Alright, in this case I _will_ be changing it, thanks for the heads up.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

**Comment by mrtracy** [comment]
Review status: 6 of 13 files reviewed at latest revision, 1 unresolved discussion.

---

*[server/status/status.proto, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDHaIDma-CLb8nENKKN:-KDKYrTiR5ZHip9cUarX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2226011332ac5969882f4dcbb29d1848c43c1832/server/status/status.proto#L32)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

**Comment by tamird** [comment]
We should talk about @bramgruneir's point - seems reasonable to "properly" include all the metrics in the proto, since the set of all metrics recorded by a given version of CR is known statically, and adding/removing fields from a proto is at least as safe as adding/removing map keys.

---

Reviewed 1 of 13 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[server/node_test.go, line 330 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDK_tmZIFKq3PrpHrIN:-KDK_tmZIFKq3PrpHrIO:-1002498365) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/node_test.go#L330)):*
the values in this map don't seem to matter - use a set?

---

*[server/node_test.go, line 564 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDKaTokJr_HF0xGXLAJ:-KDKaTokJr_HF0xGXLAK:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/node_test.go#L564)):*
?

---

*[server/status/recorder.go, line 294 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDKai6igi4nsEq_pf6P:-KDKai6igi4nsEq_pf6Q:2056447228) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/status/recorder.go#L294)):*
this should be %T

---

*[server/status/recorder.go, line 312 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDKamChx0LEt5ZZU4Wr:-KDKamChx0LEt5ZZU4Ws:-948059205) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/status/recorder.go#L312)):*
s/Warningf/Warning/ and remove the call to `Error()`

---

*[server/status/status.proto, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDK_hB02H9sc1tDYSru:-KDK_hB02H9sc1tDYSrv:1386491476) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/status/status.proto#L36)):*
this should be nullable=false

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-:-KDKZeHAekSJnD6KnBD_:1220135008)*

**Comment by mrtracy** [comment]
Review status: 11 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[server/node_test.go, line 330 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDK_tmZIFKq3PrpHrIN:-KDLjQm-hQ2Ezr-Ub2Cp:-675021261) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/node_test.go#L330)):*
The values are used later in the test.

---

*[server/node_test.go, line 564 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDKaTokJr_HF0xGXLAJ:-KDLiyd7RJviCwdh578U:-1214456402) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/node_test.go#L564)):*
Almost in disbelief that there isn't a linter for this!  Removed.

---

*[server/status/recorder.go, line 294 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDKai6igi4nsEq_pf6P:-KDLjZ9NnAV0Q87FX7fq:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/status/recorder.go#L294)):*
Done.

---

*[server/status/recorder.go, line 312 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDKamChx0LEt5ZZU4Wr:-KDLjnoZLgFkPFXuy9si:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/status/recorder.go#L312)):*
Done.

---

*[server/status/status.proto, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDK_hB02H9sc1tDYSru:-KDLk4EBodIZdB1A_6Ab:855277527) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/status/status.proto#L36)):*
The map or the list?  What does nullable=false do for a collection?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

**Comment by tamird** [comment]
Reviewed 3 of 7 files at r3.
Review status: 14 of 18 files reviewed at latest revision, 1 unresolved discussion.

---

*[server/status/status.proto, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDK_hB02H9sc1tDYSru:-KDLl4b1zKu4N8DoJ8aF:1605763875) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/status/status.proto#L36)):*
It makes it a collection of values, rather than pointers.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

**Comment by mrtracy** [comment]
Typescript work is complete, this is ready for merge (after code review is completed).

**Comment by tamird** [comment]
Review status: 14 of 18 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[server/status/status.proto, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDK_hB02H9sc1tDYSru:-KDOYkd2vulQFWzSVoum:-512415009) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/status/status.proto#L36)):*
Ping, this is not yet addressed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

**Comment by mrtracy** [comment]
Review status: 8 of 18 files reviewed at latest revision, 1 unresolved discussion.

---

*[server/status/status.proto, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDK_hB02H9sc1tDYSru:-KDPbXtSe-AKmtFmDnG3:-509583956) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/status/status.proto#L36)):*
Hey man, everyone has to sleep!  Field is now nullable.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

**Comment by tamird** [comment]
Reviewed 6 of 8 files at r4.
Review status: 14 of 18 files reviewed at latest revision, 1 unresolved discussion.

---

*[server/status/status.proto, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDK_hB02H9sc1tDYSru:-KDPbufsJXSKYOiHvcaN:238972360) ([raw file](https://github.com/cockroachdb/cockroach/blob/92832718e589c3c73547f117d875ec73bba5ca25/server/status/status.proto#L36)):*
I'm guessing you mean not nullable. 

I Only pinged because you had posted the "this is ready for merge" comment. Also, sleep when you're dead!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

**Comment by cuongdo** [comment]
:lgtm: for TypeScript changes, with some minor points

---

Reviewed 2 of 7 files at r3, 2 of 8 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[ui/ts/pages/nodes.ts, line 254 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDPnL4J6DffrOtgjzwR:-KDPnL4J6DffrOtgjzwS:-146276898) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f11341f430618c05e69723b63a85cd1e24e625a/ui/ts/pages/nodes.ts#L254)):*
Should use `MetricConstants.leaderRanges` instead of string literal

---

*[ui/ts/pages/nodes.ts, line 261 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDPnTAlf-4D4t-BmZ27:-KDPnTAlf-4D4t-BmZ28:-149221371) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f11341f430618c05e69723b63a85cd1e24e625a/ui/ts/pages/nodes.ts#L261)):*
same here

---

*[ui/ts/pages/nodes.ts, line 268 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDPnUD5Y2J4rhf0FXmu:-KDPnUD5Y2J4rhf0FXmv:-149221371) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f11341f430618c05e69723b63a85cd1e24e625a/ui/ts/pages/nodes.ts#L268)):*
same here

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-:-KDPpTstNUJQ2XHB4yVy:2001384514)*

**Comment by mrtracy** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[ui/ts/pages/nodes.ts, line 254 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDPnL4J6DffrOtgjzwR:-KDQ0o3w7chEx-vVHgfh:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f11341f430618c05e69723b63a85cd1e24e625a/ui/ts/pages/nodes.ts#L254)):*
Done.

---

*[ui/ts/pages/nodes.ts, line 261 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDPnTAlf-4D4t-BmZ27:-KDQ0ph5m3srCMoofO4m:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f11341f430618c05e69723b63a85cd1e24e625a/ui/ts/pages/nodes.ts#L261)):*
Done.

---

*[ui/ts/pages/nodes.ts, line 268 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412#-KDPnUD5Y2J4rhf0FXmu:-KDQ0rJV3ZZzH7ygXmwx:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f11341f430618c05e69723b63a85cd1e24e625a/ui/ts/pages/nodes.ts#L268)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5412)*

File: ui/ts/models/proto.ts [hunk 1]
```diff
@@ -128,62 +63,82 @@ module Models {
...
+      export var valCount: string = "valcount";
+      export var intentCount: string = "intentcount";
+      export var intentAge: string = "intentage";
+      export var gcBytesAge: string = "gcbytesage";
+      export var lastUpdateNano: string = "lastupdatenanos";
+      export var capacity: string = "capacity";
+      export var availableCapacity: string = "capacity.available";
+      export var sysBytes: string = "sysbytes";
+      export var sysCount: string = "syscount";
+    }
```
Comment (maxlang) [inline ui/ts/models/proto.ts]: Any reason you decided to do it this way? Even if you represent it as a map on the backend, we lost some typescript usefulness here if we don't specify these as types on "StatusMetrics." And especially if you're enumerating them all anyway, it seems like you might as well enumerate them in an interface vs as constants.

**Pull request merged**